src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -0,0 +1,2164 @@
+        UseWinInetProxy = 2, // WPAD protocol and PAC files supported.
+        UseCustomProxy = 3 // Use the custom proxy specified in the Proxy property.
+    }
+
+#if HTTP_DLL
+    internal enum AutomaticRedirectionPolicy
+#else
+    public enum AutomaticRedirectionPolicy
+#endif
+    {
+        Never = 0,
+        DisallowHttpsToHttp = 1,
+        Always = 2
+    }
+
+#if HTTP_DLL
+    internal enum CookieUsePolicy
+#else
+    public enum CookieUsePolicy
+#endif
+    {
+        IgnoreCookies = 0,
+        UseInternalCookieStoreOnly = 1,
+        UseSpecifiedCookieContainer = 2
+    }
+
+#if HTTP_DLL
+    internal class WinHttpHandler : HttpMessageHandler
+#else
+    public class WinHttpHandler : HttpMessageHandler
+#endif
+    {
+        private static Interop.WinHttp.WINHTTP_STATUS_CALLBACK s_staticCallback = 
+            new Interop.WinHttp.WINHTTP_STATUS_CALLBACK(WinHttpStatusCallback);
+
+        // TODO:  This looks messy but it is fast. Research a cleaner way
+        // to do this which keeps high performance lookup.
+        //
+        // Fast lookup table to convert WINHTTP_AUTH constants to strings.
+        // WINHTTP_AUTH_SCHEME_BASIC = 0x00000001;
+        // WINHTTP_AUTH_SCHEME_DIGEST = 0x00000008;
+        // WINHTTP_AUTH_SCHEME_NEGOTIATE = 0x00000010;
+        private static readonly string[] s_authSchemeStringMapping =
+        {
+            null,
+            "Basic",
+            null,
+            null,
+            null,
+            null,
+            null,
+            null,
+            "Digest",
+            null,
+            null,
+            null,
+            null,
+            null,
+            null,
+            null,
+            "Negotiate"
+        };
+
+        private static readonly uint[] s_authSchemePriorityOrder =
+        {
+            Interop.WinHttp.WINHTTP_AUTH_SCHEME_NEGOTIATE,
+            Interop.WinHttp.WINHTTP_AUTH_SCHEME_DIGEST,
+            Interop.WinHttp.WINHTTP_AUTH_SCHEME_BASIC
+        };
+
+        #region Constants
+        private const string ClientAuthenticationOID = "1.3.6.1.5.5.7.3.2";
+        private const string UriSchemeHttps = "https";
+        private const string HeaderNameContentLength = "Content-Length";
+        private const string HeaderNameContentEncoding = "Content-Encoding";
+        private const string HeaderNameCookie = "Cookie";
+        private const string HeaderNameSetCookie = "Set-Cookie";
+        private const string EncodingNameDeflate = "deflate";
+        private const string EncodingNameGzip = "gzip";
+        private const string EmptyCookieHeader = HeaderNameCookie + ":";
+        private static readonly string[] s_httpHeadersSeparator = { "\r\n" };
+        private static readonly TimeSpan s_maxTimeout = TimeSpan.FromMilliseconds(int.MaxValue);
+        private static readonly TimeSpan s_infiniteTimeout = Timeout.InfiniteTimeSpan;
```

Comment:
It's not a big deal, but why do we duplicate the static readonly Timeout.InfiniteTimeSpan field into another static readonly field, rather than just using Timeout.InfiniteTimeSpan everywhere s_inifiniteTimeout is currently used?